std/experimental/color/rgb.d

```diff
@@ -0,0 +1,852 @@
+{
+    static if(chars.length == 0)
+        enum AnyIn = false;
+    else
+        enum AnyIn = CanFind!(s, chars[0]) || AnyIn!(s, chars[1..$]);
+}
+template NoneIn(string s, string chars)
+{
+    static if(chars.length == 0)
+        enum NoneIn = true;
+    else
+        enum NoneIn = !CanFind!(s, chars[0]) && NoneIn!(s, chars[1..$]);
+}
+template NotIn(string s, string chars)
+{
+    static if(chars.length == 0)
+        enum NotIn = char(0);
+    else static if(!CanFind!(s, chars[0]))
+        enum NotIn = chars[0];
+    else
+        enum NotIn = NotIn!(s, chars[1..$]);
+}
+
+unittest
+{
+    static assert(CanFind!("string", 'i'));
+    static assert(!CanFind!("string", 'x'));
+    static assert(AllIn!("string", "sgi"));
+    static assert(!AllIn!("string", "sgix"));
+    static assert(AnyIn!("string", "sx"));
+    static assert(!AnyIn!("string", "x"));
+    static assert(NoneIn!("string", "xyz"));
+    static assert(!NoneIn!("string", "sxyz"));
+}
+
+
+// shorthand alias for all components
+alias AllComponents = TypeTuple!('l','r','g','b','a');
+
+// build mixin code to perform expresions per-element
+template ComponentExpression(string expression, char component, string op)
+{
+    template BuildExpression(string e, char c, string op)
+    {
+        static if(e.length == 0)
+            enum BuildExpression = "";
+        else static if(e[0] == '_')
+            enum BuildExpression = c ~ BuildExpression!(e[1..$], c, op);
+        else static if(e[0] == '#')
+            enum BuildExpression = op ~ BuildExpression!(e[1..$], c, op);
+        else
+            enum BuildExpression = e[0] ~ BuildExpression!(e[1..$], c, op);
+    }
+    enum ComponentExpression =
+        "static if(CanFind!(components, '"~component~"'))" ~ "\n\t" ~
+            BuildExpression!(expression, component, op);
+}
+
+
+// try and use the preferred float type, but if the int type exceeds the preferred float precision, we'll upgrade the float
+template FloatTypeFor(IntType, RequestedFloat = float)
+{
+    static if(IntType.sizeof > 2)
+        alias FloatTypeFor = double;
+    else
+        alias FloatTypeFor = RequestedFloat;
+}
+
+// find the fastest type to do format conversion without losing precision
+template WorkingType(From, To)
+{
+    static if(isIntegral!From && isIntegral!To)
+    {
+        // small integer types can use float and not lose precision
+        static if(From.sizeof <= 2 && To.sizeof <= 2)
+            alias WorkingType = float;
+        else
+            alias WorkingType = double;
+    }
+    else static if(isIntegral!From && isFloatingPoint!To)
+        alias WorkingType = To;
+    else static if(isFloatingPoint!From && isIntegral!To)
+        alias WorkingType = FloatTypeFor!To;
+    else
+    {
+        static if(From.sizeof > To.sizeof)
+            alias WorkingType = From;
+        else
+            alias WorkingType = To;
+    }
+}
+
+enum isValidComponentType(T) = isIntegral!T || isFloatingPoint!T;
```

Comment:
I think that conversion from colour to ascii would fall under the umbrella of 'composition' ;)
That said, I wonder if I should just remove this restriction? It will cause some havoc though, lots of functions won't work with other types.